File: Iteration2/src/Domain/Course.java
package Domain;

import java.util.ArrayList;
import java.util.Collection;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
@JsonIgnoreProperties(ignoreUnknown = true)
public class Course implements Searchable{

    private String courseName;
    private String courseCode;
    private Collection<String> lecturersNumbers;
    private ArrayList<CourseSection> courseSection;
    private Collection<String> studentNumbers;
    private Collection<String> prerequisitesIds;
    private int credit;
    private int semester;

    public Course(String courseName, String courseCode, Collection<String> lecturersNumbers, ArrayList<CourseSection> courseSection, Collection<String> studentNumbers, Collection<String> prerequisitesIds, int credit) {
        this.credit = credit;
        this.courseName = courseName;
        this.courseCode = courseCode;
        this.lecturersNumbers = lecturersNumbers;
        this.courseSection = courseSection;
        this.studentNumbers = studentNumbers;
        this.prerequisitesIds = prerequisitesIds;
    }

    public Course(){};

    public void AddSection(CourseSection section){
        if(courseSection == null)
            courseSection = new ArrayList<CourseSection>();
        courseSection.add(section);
    }

    public void printCourseSections(){
        for(CourseSection each : courseSection){
            each.printSections();
        }
    }






    public int getCredit() {
        return credit;
    }

    public void setCredit(int credit) {
        this.credit = credit;
    }
    public String getCourseName() {
        return courseName;
    }

    public void setCourseName(String courseName) {
        this.courseName = courseName;
    }
    public String getCourseCode() {
        return courseCode;
    }

    public void setCourseCode(String courseCode) {
        this.courseCode = courseCode;
    }

    public Collection<String> getLecturersNumbers() {
        return lecturersNumbers;
    }

    public void setLecturersNumbers(Collection<String> lecturersIds) {
        this.lecturersNumbers = lecturersIds;
    }

    public ArrayList<CourseSection> getCourseSection() {
        return courseSection;
    }

    public void setCourseSection(ArrayList<CourseSection> courseSection) {
        this.courseSection = courseSection;
    }

    public Collection<String> getStudentNumbers() {
        return studentNumbers;
    }

    @JsonIgnore
    public Collection<Student> getStudents(){
        ArrayList<Student> students = new ArrayList<>();
        ArrayList<Student> allStudents = (ArrayList<Student>) Department.getInstance().getAllStudents();

        for(Student student : allStudents){
            if(this.getStudentNumbers().contains(student.getStudentNo())){
                students.add(student);
            }
        }
        return students;
    }

    public void setStudentNumbers(Collection<String> studentNumbers) {
        this.studentNumbers = studentNumbers;
    }

    public Collection<String> getPrerequisitesIds() {
        return prerequisitesIds;
    }

    public void setPrerequisitesIds(Collection<String> prerequisitesIds) {
        this.prerequisitesIds = prerequisitesIds;
    }

    public int getSemester() {
        return semester;
    }

    public void setSemester(int semester) {
        this.semester = semester;
    }

    @Override
    public String toString() {
        return "Course{" +
                "courseName='" + courseName + '\'' +
                ", courseCode='" + courseCode + '\'' +
                ", lecturersNumbers=" + lecturersNumbers +
                ", studentNumbers=" + studentNumbers +
                ", prerequisitesIds=" + prerequisitesIds +
                ", credit=" + credit +
                ", semester=" + semester +
                '}';
    }
}

File: Iteration3/Domain/Course.py
 # Assuming you have a Department class in your project


class Course:
    def __init__(self, course_name, course_code, lecturers_numbers, course_sections, student_numbers, prerequisites_ids, credit, semester):
        self.credit = credit
        self.course_name = course_name
        self.course_code = course_code
        self.lecturers_numbers = lecturers_numbers
        self.course_sections = course_sections
        self.student_numbers = student_numbers
        self.prerequisites_ids = prerequisites_ids
        self.semester = semester  # Assuming the default value for semester is 0

    # region Methods

    def add_section(self, section):
        if self.course_sections is None:
            self.course_sections = []
        self.course_sections.append(section)

    def print_course_sections(self):
        for each in self.course_sections:
            each.print_sections()

    def get_students(self):
        from Domain.Department import Department
        students = []
        all_students = Department.get_instance().get_all_students()

        for student in all_students:
            if student.get_student_no() in self.get_student_numbers():
                students.append(student)
        return students

    # endregion Methods

    # region Properties

    def get_credit(self):
        return self.credit

    def set_credit(self, credit):
        self.credit = credit

    def get_course_name(self):
        return self.course_name

    def set_course_name(self, course_name):
        self.course_name = course_name

    def get_course_code(self):
        return self.course_code

    def set_course_code(self, course_code):
        self.course_code = course_code

    def get_lecturers_numbers(self):
        return self.lecturers_numbers

    def set_lecturers_numbers(self, lecturers_ids):
        self.lecturers_numbers = lecturers_ids

    def get_course_sections(self):
        return self.course_sections

    def set_course_sections(self, course_sections):
        self.course_sections = course_sections

    def get_student_numbers(self):
        return self.student_numbers

    def set_student_numbers(self, student_numbers):
        self.student_numbers = student_numbers

    def get_prerequisites_ids(self):
        return self.prerequisites_ids

    def set_prerequisites_ids(self, prerequisites_ids):
        self.prerequisites_ids = prerequisites_ids

    def get_semester(self):
        return self.semester

    def set_semester(self, semester):
        self.semester = semester

    # endregion Properties

    def __str__(self):
        return f"Course{{course_name='{self.course_name}', course_code='{self.course_code}', " \
               f"lecturers_numbers={self.lecturers_numbers}, student_numbers={self.student_numbers}, " \
               f"prerequisites_ids={self.prerequisites_ids}, credit={self.credit}, semester={self.semester}}}"


import json
from Domain.Course import Course
from Domain.CourseSection import CourseSectionEncoder

class CourseEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Course):
            return {
                'credit': obj.get_credit(),
                'course_name': obj.get_course_name(),
                'course_code': obj.get_course_code(),
                'lecturers_numbers': obj.get_lecturers_numbers(),
                'course_section': [CourseSectionEncoder().default(section) for section in obj.get_course_sections()],
                'student_numbers': obj.get_student_numbers(),
                'prerequisites_ids': obj.get_prerequisites_ids(),
                'semester': obj.get_semester()
            }
        return super().default(obj)

